File: 4_step_4_binary_search/06_least_weight_capacity_to_ship_packages_in_d_days.cpp
#include <bits/stdc++.h>
using namespace std;

int getMaxElement(vector<int> &weights);
int getTotalWeight(vector<int> &weights);
bool possibility(vector<int> &weights, int capacity, int days);
int leastWeightCapacityToShipPackages(vector<int> &weights, int days);

int main()
{
    // vector<int> weights = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    // int days = 5;

    // vector<int> weights = {3, 2, 2, 4, 1, 4};
    // int days = 3;

    vector<int> weights = {1,2,3,1,1};
    int days = 4;

    int result = leastWeightCapacityToShipPackages(weights, days);
    cout << result << endl;

    return 0;
}

int getMaxElement(vector<int> &weights)
{
    int maxElement = weights[0];

    for (int i = 0; i < weights.size(); i++)
    {
        if (weights[i] > maxElement)
        {
            maxElement = weights[i];
        }
    }

    return maxElement;
}

int getTotalWeight(vector<int> &weights)
{
    int totalWeight = 0;

    for (int i = 0; i < weights.size(); i++)
    {
        totalWeight += weights[i];
    }

    return totalWeight;
}

bool possibility(vector<int> &weights, int capacity, int days)
{
    int d = 1;
    long long sum = 0;

    for (int w : weights)
    {
        if (sum + w <= capacity)
        {
            sum += w;
        }
        else
        {
            d++;
            sum = w;
        }
    }
    return d <= days;
}

int leastWeightCapacityToShipPackages(vector<int> &weights, int days)
{
    int low = getMaxElement(weights);
    int high = getTotalWeight(weights);
    int ans = -1;

    while (low <= high)
    {
        int mid = low + (high - low) / 2;

        if (possibility(weights, mid, days))
        {
            ans = mid;
            high = mid - 1;
        }
        else
        {
            low = mid + 1;
        }
    }

    return ans;
}

File: 4_step_4_binary_search/06_least_weight_capacity_to_ship_packages_in_d_days.js
function getMaxElement(weights) {
  let maxElement = weights[0];

  for (let i = 0; i<weights.length; i++) {
    if (weights[i] > maxElement) {
      maxElement = weights[i];
    }
  }

  return maxElement;
}

function getTotalWeight(weights) {
  let totalWeight = 0;

  for (let i = 0; i < weights.length; i++) {
    totalWeight += weights[i];
  }

  return totalWeight;
}

function possibility(weights, capacity, totalDays) {
  let days = 1;
  let sum = 0;

  for (let w of weights) {
    if (sum + w <= capacity) {
      sum += w;
    } else {
      days++;
      sum = w;
    }
  }

  return days <= totalDays;
}

function leastWeightCapacityToShipPackages(weights, totalDays) {
  let low = getMaxElement(weights);
  let high = getTotalWeight(weights);
  let ans = -1;

  while (low <= high) {
    let mid = Math.floor(low + (high - low) / 2);

    if (possibility(weights, mid, totalDays)) {
      ans = mid;
      high = mid - 1;
    } else {
      low = mid + 1;
    }
  }

  return ans;
}

const weights = [1,2,3,4,5,6,7,8,9,10]
const totalDays = 5
let result = leastWeightCapacityToShipPackages(weights, totalDays)
console.log(result);
